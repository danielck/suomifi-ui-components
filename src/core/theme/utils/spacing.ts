import { SpacingDesignTokens } from 'suomifi-design-tokens';
import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';
import { css } from 'styled-components';

type SpacingDesignTokensKeys = keyof SpacingDesignTokens;

export const spacingTokensKeys = Object.keys(
  defaultSuomifiTheme.spacing,
) as SpacingDesignTokensKeys[];

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

/**
 * Spacing type
 * @typedef {String} spacingType
 * @param {String} type 'padding' or 'margin'
 */
/**
 * Spacing tokens
 * @typedef {String} spacingTokens
 * @param {String} t token for top, or top and bottom, or all
 * @param {String} r token for right, or right and left
 * @param {String} b token for bottom
 * @param {String} l token for left
 */
/**
 * Set margin or padding pased on tokens
 * @param {Object} theme suomifiTheme
 * @return {(spacingType) => (spacingTokens) => String}
 */
const space =
  (theme: SuomifiTheme) =>
  (type: 'padding' | 'margin') =>
  (t?: SpacingProp, r?: SpacingProp, b?: SpacingProp, l?: SpacingProp) =>
    !!t
      ? `${!!t ? `${type}-top: ${spaceVal(theme)(t)};` : ''}
    ${!!r ? `${type}-right: ${spaceVal(theme)(r)};` : ''}
    ${!!b ? `${type}-bottom: ${spaceVal(theme)(b)};` : ''}
    ${!!l ? `${type}-left:${spaceVal(theme)(l)};` : ''}`
      : '';

/**
 * Set margin based on theme
 * @param {Object} theme
 */
export const margin = (theme: SuomifiTheme) => space(theme)('margin');
/**
 * Set padding based on theme
 * @param {Object} theme
 */
export const padding = (theme: SuomifiTheme) => space(theme)('padding');

/**
 * Create spacing styles for CSS-selector (-xxs, -xs, -s...)
 * TODO: this should be in suomifiTheme? or integrated somehow with that
 * @param {Object} tokens Design tokens
 * @return {(spacingType) => (selector: String) => String}
 */
export const spacingModifiers =
  (theme: SuomifiTheme) =>
  (
    spacing:
      | 'padding'
      | 'padding-top'
      | 'padding-right'
      | 'padding-bottom'
      | 'padding-left'
      | 'margin'
      | 'margin-top'
      | 'margin-right'
      | 'margin-bottom'
      | 'margin-left',
  ) =>
  (selector: string) =>
    spacingTokensKeys.reduce(
      (ret, k) =>
        `${ret}${selector}-${k}{
        ${spacing}: ${spaceVal(theme)(k)}
  } `,
      '',
    );

export interface SpacingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export const spacingClasses = (theme: SuomifiTheme) => css`
  ${spacingModifiers(theme)('margin')('&.fi-margin')}
  ${spacingModifiers(theme)('margin-top')('&.fi-margin-top')}
  ${spacingModifiers(theme)('margin-right')('&.fi-margin-right')}
  ${spacingModifiers(theme)('margin-bottom')('&.fi-margin-bottom')}
  ${spacingModifiers(theme)('margin-left')('&.fi-margin-left')}
  ${spacingModifiers(theme)('padding')('&.fi-padding')}
  ${spacingModifiers(theme)('padding-top')('&.fi-padding-top')}
  ${spacingModifiers(theme)('padding-right')('&.fi-padding-right')}
  ${spacingModifiers(theme)('padding-bottom')('&.fi-padding-bottom')}
  ${spacingModifiers(theme)('padding-left')('&.fi-padding-left')}
`;

export const getSpacingClassNamesFromProps = (props: SpacingProps) => ({
  [`fi-padding-${props.padding}`]: !!props.padding,
  [`fi-margin-${props.margin}`]: !!props.margin,
  [`fi-margin-top-${props.mt}`]: !!props.mt,
  [`fi-margin-right-${props.mr}`]: !!props.mr,
  [`fi-margin-bottom-${props.mb}`]: !!props.mb,
  [`fi-margin-left-${props.ml}`]: !!props.ml,
  [`fi-margin-left-${props.mx}`]: !!props.mx,
  [`fi-margin-right-${props.mx}`]: !!props.mx,
  [`fi-margin-top-${props.my}`]: !!props.my,
  [`fi-margin-bottom-${props.my}`]: !!props.my,
  [`fi-padding-top-${props.pt}`]: !!props.pt,
  [`fi-padding-right-${props.pr}`]: !!props.pr,
  [`fi-padding-bottom-${props.pb}`]: !!props.pb,
  [`fi-padding-left-${props.pl}`]: !!props.pl,
  [`fi-padding-left-${props.px}`]: !!props.px,
  [`fi-padding-right-${props.px}`]: !!props.px,
  [`fi-padding-top-${props.py}`]: !!props.py,
  [`fi-padding-bottom-${props.py}`]: !!props.py,
});

export const separateSpacingProps = (props: any) => {
  const {
    margin: marginProp,
    padding: paddingProp,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      margin: marginProp,
      padding: paddingProp,
      my,
      mx,
      py,
      px,
      mt,
      mr,
      mb,
      ml,
      pt,
      pr,
      pb,
      pl,
    },
    otherProps,
  ];
};
