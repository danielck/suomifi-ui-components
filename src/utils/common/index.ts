export {
  windowAvailable,
  useEnhancedEffect,
  getOwnerDocument,
  escapeStringRegexp,
  forkRefs,
  autocompleteTimeString,
} from './common';
