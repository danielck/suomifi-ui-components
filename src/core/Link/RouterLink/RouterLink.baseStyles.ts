import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { font } from '../../theme/reset';
import { allStates } from '../../../utils/css';

export const RouterLinkStyles = (theme: SuomifiTheme) => css`
  ${font(theme)('bodyText')}

  &.fi-link--router {
    ${allStates(`color: ${theme.colors.highlightBase};`)}
    color: ${theme.colors.highlightBase};
    text-decoration: none;

    &:focus,
    &:focus-within {
      text-decoration: none;
    }

    &:focus {
      ${theme.focus.boxShadowFocus}
    }

    &:hover,
    &:active {
      text-decoration: underline;
    }
    &:visited {
      color: ${theme.colors.accentTertiaryDark1};
    }
  }
`;
