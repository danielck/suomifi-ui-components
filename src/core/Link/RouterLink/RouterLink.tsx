// import React, { Component, ReactNode } from 'react';
import React, { ReactNode } from 'react';
import { default as styled } from 'styled-components';
import { RouterLinkStyles } from './RouterLink.baseStyles';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../../theme';
import { baseClassName } from '../BaseLink/BaseLink';
import classnames from 'classnames';
import { HtmlA } from '../../../reset';

//
// Source: https://www.benmvp.com/blog/polymorphic-react-components-typescript/
//
// Source: https://github.com/emotion-js/emotion/blob/master/packages/styled-base/types/helper.d.ts
// A more precise version of just React.ComponentPropsWithoutRef on its own
export type PropsOf<
  C extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>,
> = JSX.LibraryManagedAttributes<C, React.ComponentPropsWithoutRef<C>>;

type AsProp<C extends React.ElementType> = {
  /**
   * An override of the default HTML tag.
   * Can also be another React component.
   * @default <a>
   */
  asComponent?: C;
};

/**
 * Allows for extending a set of props (`ExtendedProps`) by an overriding set of props
 * (`OverrideProps`), ensuring that any duplicates are overridden by the overriding
 * set of props.
 */
export type ExtendableProps<
  ExtendedProps = {},
  OverrideProps = {},
> = OverrideProps & Omit<ExtendedProps, keyof OverrideProps>;

/**
 * Allows for inheriting the props from the specified element type so that
 * props like children, className & style work, as well as element-specific
 * attributes like aria roles. The component (`C`) must be passed in.
 */
export type InheritableElementProps<
  C extends React.ElementType,
  Props = {},
> = ExtendableProps<PropsOf<C>, Props>;

/**
 * A more sophisticated version of `InheritableElementProps` where
 * the passed in `as` prop will determine which props can be included
 */
export type PolymorphicComponentProps<
  C extends React.ElementType,
  Props = {},
> = InheritableElementProps<C, Props & AsProp<C>>;

//
//
//

const routerLinkClassName = `${baseClassName}--router`;

interface Props {
  /** Custom classname to extend or customize */
  className?: string;
  /**
   * Link element displayed content
   */
  children: ReactNode;
}

export type RouterLinkProps<C extends React.ElementType> =
  PolymorphicComponentProps<C, Props>;

const PolymorphicLink = <C extends React.ElementType>(
  props: RouterLinkProps<C> & SuomifiThemeProp,
) => {
  const { asComponent, children, className, theme, ...passProps } = props;
  const Component = asComponent || HtmlA;

  return (
    <Component
      className={classnames(routerLinkClassName, className)}
      {...passProps}
    >
      {children}
    </Component>
  );
};

const StyledRouterLink = styled(PolymorphicLink)`
  ${({ theme }) => RouterLinkStyles(theme)}
`;

const RouterLink = <C extends React.ElementType = 'a'>(
  props: RouterLinkProps<C>,
) => (
  <SuomifiThemeConsumer>
    {({ suomifiTheme }) => <StyledRouterLink theme={suomifiTheme} {...props} />}
  </SuomifiThemeConsumer>
);

RouterLink.displayName = 'RouterLink';
export { RouterLink };
