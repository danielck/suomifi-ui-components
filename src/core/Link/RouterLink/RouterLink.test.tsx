import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test';

import { RouterLink } from './RouterLink';

const TestRouterLink = (
  <RouterLink href="/" data-testid="test-link">
    Hey this is a test
  </RouterLink>
);

const RouterLinkAsButton = (
  <RouterLink asComponent="button">Router link as a button</RouterLink>
);

test('calling render with the same component on the same container does not remount', () => {
  const { getByTestId, container } = render(TestRouterLink);
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('test-link').textContent).toBe('Hey this is a test');
});

it('should still have the correct styles applied when rendered as something other than <a> ', () => {
  const { container } = render(RouterLinkAsButton);
  expect(container.firstChild).toHaveClass('fi-link--router');
});

test('should not have basic accessibility issues', axeTest(TestRouterLink));
