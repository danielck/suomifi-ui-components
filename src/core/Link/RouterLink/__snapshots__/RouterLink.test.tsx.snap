// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c0 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c0.fi-link--router {
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0.fi-link--router:hover,
.c0.fi-link--router:active,
.c0.fi-link--router:focus,
.c0.fi-link--router:focus-within {
  color: hsl(212,63%,45%);
}

.c0.fi-link--router:focus,
.c0.fi-link--router:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0.fi-link--router:focus {
  outline: 0;
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
}

.c0.fi-link--router:hover,
.c0.fi-link--router:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0.fi-link--router:visited {
  color: hsl(284,36%,45%);
}

<a
  class="fi-link--router c0"
  data-testid="test-link"
  href="/"
>
  Hey this is a test
</a>
`;
